### ISSUE TITLE
Keep exact bitmap data when saving to vector file

### ISSUE THREAD

**Issue by Synxis** [matplotlib/matplotlib#16268]
I sometimes annotate plots using images (png format), and I want the image data not to be touched when saving to pdf: just put the data and the correct transformation in the pdf, if possible.
So instead of having a "fixed" dpi (which does not really make sense for vector formats anyway), passing something like `dpi ='raw'` (for example) would simply do what I explained before.

Even when explicitely adding `resample = False`, matplotlib still resample the image (I know this value is the default, but still). I don't know if this is a bug report or a feature request, I'm enclined to the former.

```python
import numpy as np
import matplotlib
matplotlib.use("agg")
import matplotlib.pyplot as plt

def add_image(axe, filename, position, zoom):
    img = plt.imread(filename)
    off_img = matplotlib.offsetbox.OffsetImage(img, zoom = zoom, resample = False)
    art = matplotlib.offsetbox.AnnotationBbox(off_img, position, xybox = (0, 0),
        xycoords = axe.transAxes, boxcoords = "offset points", frameon = False)
    axe.add_artist(art)

# ==========
fig = plt.figure()
axe = plt.axes()
fig.set_size_inches(3, 1.5)
axe.plot(np.arange(10), np.arange(10))

add_image(axe, "image.png", position = (0.2, 0.7), zoom = 0.07)

fig.savefig("temp.pdf", bbox_inches = "tight", pad_inches = 0)
```

Use it with [this image from wikipedia](https://upload.wikimedia.org/wikipedia/commons/4/47/PNG_transparency_demonstration_1.png) as `image.png`.

**Actual outcome (screenshot):**

![actual][2]

**Expected outcome (screenshot):**

![expected][1]

  [1]: https://i.stack.imgur.com/1f1Ro.png
  [2]: https://i.stack.imgur.com/Ekh5x.png

**Matplotlib:**
 - Version: 3.1.2
 - Backend: PDF
 - Python: 3.7 64bit
 - Installed via: `pip`
 - OS: Windows 10

**Comment by anntzer** [comment]
Your example is not reproductible, because we do not have access to image.png.

**Comment by anntzer** [comment]
Passing e.g. `dpi=400` to savefig appears to workaround the issue, though a cleaner solution would probably be better...

**Comment by anntzer** [comment]
Sorry about the incorrect message re: missing link, I did not read the issue carefully.  I agree that not having to resample would be better.

**Comment by jklymak** [comment]
You don’t have to resample, but what is the API for knowing what dpi the user wants the image to be displayed at in the PDF?  Or if you prefer, how big it will be?

**Comment by ImportanceOfBeingErnest** [comment]
Matplotlib provides `figimage` which does not resample anything. But it seems, that is not actually desired here. Because the use of an offset box with a `zoom` implicitely means that something needs to be resampled, else how, and with respect to what, would `zoom` be definable?

**Comment by Synxis** [comment]
I argue that offset & zoom are strictly for the positioning and scaling of the image but do not give any information about the resolution of the image. This would be like placing an image in svg, where you specify the position and size, but the resolution of the image is untouched.

IMHO it would be better to explicitely separate the transformation (where to place the element) and the resolution (the data to show).

Maybe `figimage` would work, I'll try it, its problem is that the coordinates are relative to the figure, not the axes. That means, moving / changing the axes would incur a lot of adjustments regarding the position of the image.

**Comment by ImportanceOfBeingErnest** [comment]
Ok, I think that makes sense somehow. Looking a bit into this, one finds that in principle the images allow for non-resampled drawing. However, this cannot be steered from the outside. 
I also find the difference between `unsampled` and `resample` arguments in the image classes utterly confusing. So `unsampled` is what we would need here and it could be easily used via the following monkey-patch:

```
from matplotlib.backends.backend_mixed import MixedModeRenderer
   
def _check_unsampled_image(self, renderer):
    if isinstance(renderer, MixedModeRenderer):
        return True
    else:
        return False

matplotlib.image.BboxImage._check_unsampled_image = _check_unsampled_image
```

Just adding the above to the code from OP results in the desired image. The assumption being that once a mixed mode renderer is used (e.g. for pdf or svg output), one wants an unsampled image.

**Comment by anntzer** [comment]
Seems reasonable, I guess this could even be put into the base `_ImageBase._check_unsampled_image`?

**Comment by ImportanceOfBeingErnest** [comment]
Oh well, I don't think hardcoding this is a reasonable solution here. It would again not allow users to steer the behaviour. Also, if we change it now, it might suddenly lead to unreasonably large files for users of existing code.  
Instead, one would need to understand why there are two parameters `resample` and `unsampled` and come to a conclusion about which one is meant to do what exactly - then possibly exposing one or both with reasonable defaults.

**Comment by Synxis** [comment]
Thank you, it solves the problem!

I wonder if user control can be achieved simply by adding an option to the `dpi` option of the renderer, ie `dpi == "raw"`. This would not be the first option to this option (there is already `"figure"`), and would allow to never resample an image if explicitely specified while still retaining backward compatibility. Obviously I'm pretty sure there is a better name for that ;)

This would mean adding this to `matplotlib/image.py`:

```python
from matplotlib.backends.backend_mixed import MixedModeRenderer

class _ImageBase(...):

    # ...

    def _check_unsampled_image(self, renderer):
        return isinstance(renderer, MixedModeRenderer) and renderer.dpi == "raw"
        # instead of  return False
```

**Comment by jklymak** [comment]
I don’t understand how the PDF renderer knows how large to amke the image?

**Comment by ImportanceOfBeingErnest** [comment]
Yeah, that confused me as well. As it stands, it's just the number of pixels in units of points. E.g. a 100 x 100 pixel image would be 100 x 100 points large. (In the case here, multiplied by the `zoom`.)
Arguably that definition is a bit arbitrary, but not completely nonsensical.


<hr>

Btw. here is how an unsampled image of just a few pixels would look in Inkscape using the hack from above:

![image](https://user-images.githubusercontent.com/23121882/72773785-8ced7a80-3c08-11ea-96ac-c4967964ae4d.png)

Compare to the current standard:

![image](https://user-images.githubusercontent.com/23121882/72773813-a4c4fe80-3c08-11ea-9873-ec752263e86f.png)

This tells us, that we absolutely need some user interface and probably should not change the default.

**Comment by jklymak** [comment]
So for the above case if the figure dpi is 200 the size of the image is still in points?   Zoom is relative to points or the current dpi?  (Sorry happy to test myself when I get time but if it’s easy for you to check).

**Comment by anntzer** [comment]
re: svg: I think this is controllable, see https://github.com/matplotlib/matplotlib/issues/10112.
re: positioning: I had to reverse-engineer some of that for mplcairo but it's indeed a bit of a mess and for example I never figured out how to make figimage() ("non-resampled") work for pdf output.  See also https://github.com/matplotlib/mplcairo/issues/19.

**Comment by cover-me** [comment]
Why not create a frameless ax and use imshow(interpolation='none') if you don't need to rotate the image? That will give you unsampled images in PDF or SVG files.

**Comment by alhom** [comment]
Resurrecting this issue a bit, seeing this is still open.

My use case for this issue is that I would like to have raster plots (presented with imshow), at their native resolutions, in an .svg output file, at two different levels: I have insets of data for potentially each pixel in the top-level imshow plot, and I'd like to present these insets at native resolution in the background context - so using inset_axes is reasonably straightforward to place the insets with top-level data coordinates.

As it is, all imshow plots are forcibly resampled to the top-level figure dpi, which is not great when I'd want some 5000 dpi at top level to have the insets properly legible (and preferably more). As it happens, I can do that! The imshow background exported to svg is not upsampled, and I can get crispy-ish insets.

Maxing out the dpi is not a really good solution, though. For example, Agg is readily complaining about 2^16 pixels per dimension max (or segfaulting). So, I guess the backend is trying to expand the whole to-be-vector figure still to a raster, despite there being no rasters covering the whole figure?

**Comment by jklymak** [comment]
For vector backends, there is currently just one DPI for any rasters in the figure.  Images are resampled to that dpi.  I don't understand why you would want an inset with 5000 dpi unless you are going to look at it on a monitor with 25-times zoom.  If you don't downsample your data, or you don't let Matplotlib resample the image, you will definitely have aliasing artifacts.  

This is still open because it is a different API than what we currently do for `imshow`, and one specific to just the vector backends.  I don't think its a bad idea, but it would take some design.

**Comment by alhom** [comment]
>  I don't understand why you would want an inset with 5000 dpi unless you are going to look at it on a monitor with 25-times zoom.

That is pretty much what I am doing - I'm looking at what is approximately an 500^2 raster dataset, and I associate with each pixel (representing a grid cell) a 800px^2 plot. That is in principle - in practice I inspect sparse-ish subsets. Since I still want to have e.g. a plot for each pixel along some interesting line, I like to have these insets at that scale to have them lined up properly in their context. And since the background is relatively coarse, I can do with what is essentially a very low dpi for the background (which actually works, since upsampling that is cut short, at least!), but the having the insets print out in any reasonable resolution requires having an unreasonable dpi for the whole figure, which seems to cause all sorts of problems.

**Comment by jklymak** [comment]
Perhaps we could continue conversation at discourse.matplotlib.org?  This is somewhat orthogonal to this issue.

**Comment by QuLogic** [comment]
All Artists have a `rasterized` option which rasterizes them at the vector backend DPI. What if, for images, we set that to True _by default_, but if subsequently set to `False`, it will embed at whatever resolution the source data is (assuming this all can actually be conveyed to the backend properly)?
